runconfig/opts/parse_test.go

```diff
@@ -385,11 +385,29 @@ func TestParseWithMemorySwap(t *testing.T) {
 }
 
 func TestParseHostname(t *testing.T) {
-	hostname := "--hostname=hostname"
+	validHostnames := map[string]string{
+		"hostname":    "hostname",
+		"host-name":   "host-name",
+		"hostname123": "hostname123",
+		"123hostname": "123hostname",
+	}
+	invalidHostnames := map[string]string{
+		"^hostname": "invalid hostname format for --hostname: ^hostname",
+		"hostname%": "invalid hostname format for --hostname: hostname%",
+		"host&name": "invalid hostname format for --hostname: host&name",
+		"-hostname": "invalid hostname format for --hostname: -hostname",
```

Comment:
Looking at the issue this PR is resolving, I don't see any host names containing an underscore here? Original issue was (https://github.com/docker/docker/issues/20371);

> hostname does not check for the _ character which is not allowed in hostnames.